L2 announcement (#25471)
cilium-team/cilium: Documentation/network/l2-announcements.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _l2_announcements:

************************************
L2 Announcements / L2 Aware LB
************************************

L2 Announcements is a feature which makes services visible and reachable on 

> the local area network. This feature is primarily intended for bare metal 

Reviewer: ```suggestion
the local area network. This feature is primarily intended for on-premises
```
Isn't it more about on-prem (where it's easy achievable to announce a specific IP via ARP) vs. public cloud, rather than bare metal vs. something? Also, traditional VMs (e.g. hosted on ESXi) are probably even more common than real bare metal nodes.

Author: Yes, very true, thx!

campus networks.

When used, this feature will respond to ARP queries for ExternalIPs and/or 
LoadBalancer IPs. These IPs are Virtual IPs (not installed on network 
devices) on multiple nodes, so for each service one node at a time will respond
to the ARP queries and respond with its MAC address. This node will perform 
load balancing with the service load balancing feature, thus acting as a 
north/south load balancer.

The advantage over this feature over NodePort services is that each service can
use a unique IP so multiple service can use the same port numbers. When using 
NodePorts its up to the client to decide to which host to send traffic, if a node
goes down, the IP+Port combo becomes unusable. With L2 announcements the service
VIP simply migrates to another node and will continue to work.

The L2 Announcements feature and all the requirements can be enabled as follow:

.. tabs::
    .. group-tab:: Helm

        .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
               --namespace kube-system \\
               --reuse-values \\
               --set l2announcements.enabled=true \\

>                --set kubeProxyReplacement=strict

Reviewer: This is a non-functional `helm` example. Should we add the following arguments like we do in the [KPR guide](https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#quick-start)?
```
    --set k8sServiceHost=${API_SERVER_IP} \
    --set k8sServicePort=${API_SERVER_PORT}
```

Author: Ah, yes, we should. I copied the Egress gateway example and did not notice this.

    .. group-tab:: ConfigMap

        .. code-block:: yaml

            enable-l2-announcements: true
            kube-proxy-replacement: strict

Prerequisites
#############

> * Kube Proxy replacement mode must be enabled, see :ref:`kubeproxy-free` for details.

Reviewer: ```suggestion
* Kube Proxy replacement mode must be enabled and set to `strict` mode, see :ref:`kubeproxy-free` for details.
```
I think it would make sense to be precise here. What does "enabled" regarding KPR mean? Is `partial` also fine?

Author: I think it also works in some configuration with `partial` mode, but I honestly can't figure out in which combination we do and don't load+attach the `bpf_host` programs which is the main requirement. So I will keep it at strict for now.

> * All devices on which L2 Aware LB should be enabled should be included in the 

Reviewer: ```suggestion
* All devices on which L2 Aware LB should be enabled and included in the 
```

  ``--devices`` flag or ``devices`` Helm option if explicitly set, see :ref:`NodePort Devices`.

* The ``externalIPs.enable=true`` Helm option should be set, if usage of externalIPs
  is desired. Otherwise service load balancing for external IPs is disabled.

Limitations
###########

> * The feature currently does not support IPv6/NDP

Reviewer: ```suggestion
* The feature currently does not support IPv6/NDP.
```

* Due to the way L3->L2 translation protocols work, one node receives all 
  per IP, no load balancing before traffic hits the cluster.

* The feature currently has no traffic balancing mechanism so nodes within the
  same policy might be asymmetrically loaded. For details see :ref:`l2_announcements_leader_election`.

Policies
########

Policies provide fine-grained control over which services should be announced,

> where, and how. This is an example of policy using all optional fields:

Reviewer: ```suggestion
where, and how. This is an example policy using all optional fields:
```

        matchLabels:
          role: worker
      interfaces:
      - ^eth[0-9]+
      externalIPs: true
      loadBalancerIPs: true  

Service Selector
----------------

The service selector is a `label selector <https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/>`__ 
that determines which services are selected by this policy. If no service 
selector is provided, all services are selected by the policy.

There are a few special purpose selector fields which don't match on labels but
instead on other metadata like ``.meta.name`` or ``.meta.namespace``.

=============================== ===================
Selector                        Field
------------------------------- -------------------
io.kubernetes.service.namespace ``.meta.namespace``
io.kubernetes.service.name      ``.meta.name``
=============================== ===================

Node Selector
-------------

The node selector field is a `label selector <https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/>`__
which determines which nodes are candidates to announce the services from.

It might be desirable to pick a subset of nodes in you cluster, since the chosen

> node (see :ref:`l2_announcements_leader_election`) will act as the North/South

Reviewer: ```suggestion
node (see :ref:`l2_announcements_leader_election`) will act as the north/south
```
Let's unify "north/south". Above it was written without capital letters.

Interfaces
----------

The interfaces field is a list of regular expressions (`golang syntax <https://pkg.go.dev/regexp/syntax>`__
that determine over which network interfaces the selected services will be 
announced. This field is optional, if not specified all interfaces will be used.

The expressions are OR-ed together, so any network device matching any of the 
expressions will be matched.

L2 announcements only work if the selected devices are also part of the set of 
devices specified in the ``devices`` Helm option, see :ref:`NodePort Devices`

.. note::
    This selector is NOT a security feature, services will still be available 
    via interfaces when not advertised (for example by hard-coding ARP entries).

IP Types
--------

The ``externalIPs`` and ``loadBalancerIPs`` fields determine what sort of IPs 
are announced. They are ``false`` by default, so a functional policy should always
have one or both set to ``true``.

If ``externalIPs`` is ``true`` all IPs in `.spec.externalIPs <https://kubernetes.io/docs/concepts/services-networking/service/#external-ips>`__
are announced, which are managed by service authors.

> If ``loadBalancerIPs` is ``true`` all IPs in ``.status.loadbalacer.ingress``

Reviewer: ```suggestion
If ``loadBalancerIPs`` is ``true`` all IPs in ``.status.loadbalacer.ingress``
```

are announced. These can be assigned by :ref:`lb_ipam` which can be configured
by cluster admins for better control over which IPs can be allocated.

.. note::
    If a user intends to use ``externalIPs``, the ``externalIPs.enable=true`` 
    Helm option should be set to enable service load balancing for external IPs.

Status
------

If a policy is invalid for any number of reasons, the status of the policy will reflect that.

> For example if a invalid match expression is provided:

Reviewer: ```suggestion
For example if an invalid match expression is provided:
```

      resourceVersion: "449966"
      uid: e3c9c020-6e24-4c5c-9df9-d0c50f6c4cec
    spec:
      acquireTime: "2023-05-09T15:14:20.108431Z"
      holderIdentity: worker-node
      leaseDurationSeconds: 3
      leaseTransitions: 1
      renewTime: "2023-05-12T12:15:26.773020Z"

The ``acquireTime`` is the time at which the current leader acquired the lease.
The ``holderIdentity`` is the name of the current holder/leader node. 
If the leader does not renew the lease for ``leaseDurationSeconds`` seconds a
new leader is chosen. ``leaseTransitions`` indicates how often the lease changed
hands and ``renewTime`` the last time the leader renewed the lease.

There are three Helm options that can be tuned with regards to leases:

* ``l2announcements.leaseDuration`` determines the ``leaseDurationSeconds`` value
  of created leases and by extent how long a leader must be "down" before 
  failover occurs. (default is 15s)

* ``l2announcements.leaseRenewDeadline`` is the interval at which the leader 
  should renew the lease. (default is 5s)

* ``l2announcements.leaseRetryPeriod`` if renewing the lease fails, how long 
  should the agent wait before it tries again. (default is 2s)

> 

Reviewer: We should perhaps add a note regarding the allowed values ranges. E.g. can `l2announcements.leaseDuration` only be `1s` or more because this value is used for `leaseDurationSeconds`?
Also, what about `leaseRenewDeadline` and `leaseRetryPeriod `: Do both support sub-second configurations (`ms`) or only `leaseRetryPeriod ` (like you show in the `helm upgrade` example below)?

Author: I will follow up on that, have to test what actually happens when you go below 1 second, according to the checks in the go-client library its valid: https://github.com/kubernetes/client-go/blob/v0.27.1/tools/leaderelection/leaderelection.go#L77-L97

> Also, what about leaseRenewDeadline and leaseRetryPeriod : Do both support sub-second configurations (ms) or only leaseRetryPeriod  (like you show in the helm upgrade example below)?

They all should support sub second values.

Author: Having a duration lower than `leaseDurationSeconds` does not do anything, it just gets rounded to the nearest second. So I will adjust the docs and the code to check for that and to document the limit. Thanks for the sharp eyes.

.. tabs::
    .. group-tab:: Helm

        .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
               --namespace kube-system \\
               --reuse-values \\
               --set l2announcements.enabled=true \\

>                --set kubeProxyReplacement=strict \\

Reviewer: Same as above. This is basically a non-functional `helm` example. Should we add the following arguments like we do in the [KPR guide](https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#quick-start)?
```
    --set k8sServiceHost=${API_SERVER_IP} \
    --set k8sServicePort=${API_SERVER_PORT}
```

               --set l2announcements.leaseDuration=3s \\
               --set l2announcements.leaseRenewDeadline=1s \\
               --set l2announcements.leaseRetryPeriod=200ms

    .. group-tab:: ConfigMap

        .. code-block:: yaml

            enable-l2-announcements: true
            kube-proxy-replacement: strict
            l2-announcements-lease-duration: 3s
            l2-announcements-renew-deadline: 1s
            l2-announcements-retry-period: 200ms

There is a trade-off between fast failure detection and CPU + network usage. 
Each service incurs a CPU and network overhead, so clusters with smaller amounts
of services can more easily afford faster failover times. Larger clusters might
need to increase parameters if the overhead is to much.

Failover
--------

When nodes participating in leader election detect that the lease holder did not
renew the lease for ``leaseDurationSeconds`` amount of seconds, they will ask

> the API server to make them the new holder. The first request to be processed 

Reviewer: > When nodes participating in leader election detect that the lease holder did not
renew the lease for ``leaseDurationSeconds`` amount of seconds [...]

Does this detection add some "latency" in addition to `leaseDurationSeconds` for the switch-over process or is this nearly instant because of an API watch operation?

Author: If I am reading it right, `renew` calls `tryAquireOrRenew` every `RenewDeadline`: https://github.com/kubernetes/client-go/blob/v0.27.1/tools/leaderelection/leaderelection.go#L262

At that time we check if the lease expired. So in reality the failover is `leaseDurationSeconds` at best and `leaseDurationSeconds + RenewDeadline` at worst.

Reviewer: Alright, thanks for checking. Do you think it's worth mentioning this in the doc?

Author: Yea, I think it is, will do!

Reviewer: Awesome, thanks a lot.

gets through and the rest are denied.

> When a node becomes the leader/holder, it will send out a gratuitous ARP reply 

Reviewer: We should probably mention Unsolicited Neighbor Advertisements as well or at least add context that this is IPv4 only. In the limitation section we already mentioned it currently doesn't support IPv6 so perhaps let's go for the latter?

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _l2_announcements:

************************************
L2 Announcements / L2 Aware LB
************************************

L2 Announcements is a feature which makes services visible and reachable on 
the local area network. This feature is primarily intended for on-premises
deployments within networks without BGP based routing such as office or 
campus networks.

When used, this feature will respond to ARP queries for ExternalIPs and/or 
LoadBalancer IPs. These IPs are Virtual IPs (not installed on network 
devices) on multiple nodes, so for each service one node at a time will respond
to the ARP queries and respond with its MAC address. This node will perform 
load balancing with the service load balancing feature, thus acting as a 
north/south load balancer.

> The advantage over this feature over NodePort services is that each service can

Reviewer 2: ```suggestion
The advantage of this feature over NodePort services is that each service can
```

> use a unique IP so multiple service can use the same port numbers. When using 

Reviewer 2: ```suggestion
use a unique IP so multiple services can use the same port numbers. When using 
```

> NodePorts its up to the client to decide to which host to send traffic, if a node

Reviewer 2: ```suggestion
NodePorts, it is up to the client to decide to which host to send traffic, and if a node
```

goes down, the IP+Port combo becomes unusable. With L2 announcements the service
VIP simply migrates to another node and will continue to work.

> The L2 Announcements feature and all the requirements can be enabled as follow:

Reviewer 2: ```suggestion
The L2 Announcements feature and all the requirements can be enabled as follows:
```

.. tabs::
    .. group-tab:: Helm

        .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
               --namespace kube-system \\
               --reuse-values \\
               --set l2announcements.enabled=true \\
               --set kubeProxyReplacement=strict \\
               --set k8sServiceHost=${API_SERVER_IP} \\
               --set k8sServicePort=${API_SERVER_PORT}
               

    .. group-tab:: ConfigMap

        .. code-block:: yaml

            enable-l2-announcements: true
            kube-proxy-replacement: strict

Prerequisites
#############

* Kube Proxy replacement mode must be enabled and set to `strict` mode, :ref:`kubeproxy-free` for details.

> * All devices on which L2 Aware LB should be enabled and included in the 

Reviewer 2: ```suggestion
* All devices on which L2 Aware LB will be announced should be enabled and included in the 
```

  ``--devices`` flag or ``devices`` Helm option if explicitly set, see :ref:`NodePort Devices`.

> * The ``externalIPs.enable=true`` Helm option should be set, if usage of externalIPs

Reviewer 2: ```suggestion
* The ``externalIPs.enabled=true`` Helm option should be set, if usage of externalIPs
```

  is desired. Otherwise service load balancing for external IPs is disabled.

Limitations
###########

* The feature currently does not support IPv6/NDP.

* Due to the way L3->L2 translation protocols work, one node receives all 

>   per IP, no load balancing before traffic hits the cluster.

Reviewer 2: ```suggestion
* Due to the way L3->L2 translation protocols work, one node receives all 
  ARP requests for a specific IP, so no load balancing can happen before traffic hits the cluster.
```

which determines which nodes are candidates to announce the services from.

It might be desirable to pick a subset of nodes in you cluster, since the chosen
node (see :ref:`l2_announcements_leader_election`) will act as the north/south
load balancer for all of the traffic for a particular service.

Interfaces
----------

The interfaces field is a list of regular expressions (`golang syntax <https://pkg.go.dev/regexp/syntax>`__
that determine over which network interfaces the selected services will be 
announced. This field is optional, if not specified all interfaces will be used.

The expressions are OR-ed together, so any network device matching any of the 
expressions will be matched.

L2 announcements only work if the selected devices are also part of the set of 
devices specified in the ``devices`` Helm option, see :ref:`NodePort Devices`

.. note::
    This selector is NOT a security feature, services will still be available 
    via interfaces when not advertised (for example by hard-coding ARP entries).

IP Types
--------

The ``externalIPs`` and ``loadBalancerIPs`` fields determine what sort of IPs 

> are announced. They are ``false`` by default, so a functional policy should always

Reviewer 2: ```suggestion
are announced. They are both set to ``false`` by default, so a functional policy should always
```

have one or both set to ``true``.

> If ``externalIPs`` is ``true`` all IPs in `.spec.externalIPs <https://kubernetes.io/docs/concepts/services-networking/service/#external-ips>`__
> are announced, which are managed by service authors.

Reviewer 2: ```suggestion
If ``externalIPs`` is ``true`` all IPs in the service's `.spec.externalIPs <https://kubernetes.io/docs/concepts/services-networking/service/#external-ips>`__
field are announced. These IPs are are managed by service authors.
```

> If ``loadBalancerIPs`` is ``true`` all IPs in ``.status.loadbalacer.ingress``

Reviewer 2: ```suggestion
If ``loadBalancerIPs`` is ``true`` all IPs in the service's ``.status.loadbalacer.ingress`` field
```

  should the agent wait before it tries again. (default is 2s)

.. tabs::
    .. group-tab:: Helm

        .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
               --namespace kube-system \\
               --reuse-values \\
               --set l2announcements.enabled=true \\
               --set kubeProxyReplacement=strict \\
               --set k8sServiceHost=${API_SERVER_IP} \\
               --set k8sServicePort=${API_SERVER_PORT} \\
               --set l2announcements.leaseDuration=3s \\
               --set l2announcements.leaseRenewDeadline=1s \\
               --set l2announcements.leaseRetryPeriod=200ms

    .. group-tab:: ConfigMap

        .. code-block:: yaml

            enable-l2-announcements: true
            kube-proxy-replacement: strict
            l2-announcements-lease-duration: 3s
            l2-announcements-renew-deadline: 1s
            l2-announcements-retry-period: 200ms

There is a trade-off between fast failure detection and CPU + network usage. 
Each service incurs a CPU and network overhead, so clusters with smaller amounts
of services can more easily afford faster failover times. Larger clusters might

> need to increase parameters if the overhead is to much.

Reviewer 2: ```suggestion
need to increase parameters if the overhead is too high.
```

Failover
--------

When nodes participating in leader election detect that the lease holder did not
renew the lease for ``leaseDurationSeconds`` amount of seconds, they will ask
the API server to make them the new holder. The first request to be processed 
gets through and the rest are denied.

When a node becomes the leader/holder, it will send out a gratuitous ARP reply 

> over all of the configured interfaces. Clients whom accept these will update 

Reviewer 2: ```suggestion
over all of the configured interfaces. Clients who accept these will update 
```

load balancing with the service load balancing feature, thus acting as a 
north/south load balancer.

The advantage of this feature over NodePort services is that each service can
use a unique IP so multiple services can use the same port numbers. When using 
NodePorts, it is up to the client to decide to which host to send traffic, and if a node
goes down, the IP+Port combo becomes unusable. With L2 announcements the service
VIP simply migrates to another node and will continue to work.

The L2 Announcements feature and all the requirements can be enabled as follows:

.. tabs::
    .. group-tab:: Helm

        .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
               --namespace kube-system \\
               --reuse-values \\
               --set l2announcements.enabled=true \\
               --set kubeProxyReplacement=strict \\
               --set k8sServiceHost=${API_SERVER_IP} \\
               --set k8sServicePort=${API_SERVER_PORT}
               

    .. group-tab:: ConfigMap

        .. code-block:: yaml

            enable-l2-announcements: true
            kube-proxy-replacement: strict

Prerequisites
#############

> * Kube Proxy replacement mode must be enabled and set to ``strict`` mode, :ref:`kubeproxy-free` for details.

Reviewer 3: ```suggestion
* Kube Proxy replacement mode must be enabled and set to ``strict`` mode, see :ref:`kubeproxy-free` for details.
```

* All devices on which L2 Aware LB will be announced should be enabled and included in the 
  ``--devices`` flag or ``devices`` Helm option if explicitly set, see :ref:`NodePort Devices`.

* The ``externalIPs.enabled=true`` Helm option should be set, if usage of externalIPs

>   is desired. Otherwise service load balancing for external IPs is disabled.

Reviewer 3: Terminology: “should” or “must”?

(Same question for the use of “should” at other locations in the doc.)

----------------

The service selector is a `label selector <https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/>`__ 
that determines which services are selected by this policy. If no service 
selector is provided, all services are selected by the policy.

There are a few special purpose selector fields which don't match on labels but
instead on other metadata like ``.meta.name`` or ``.meta.namespace``.

=============================== ===================
Selector                        Field
------------------------------- -------------------
io.kubernetes.service.namespace ``.meta.namespace``
io.kubernetes.service.name      ``.meta.name``
=============================== ===================

Node Selector
-------------

The node selector field is a `label selector <https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/>`__
which determines which nodes are candidates to announce the services from.

It might be desirable to pick a subset of nodes in you cluster, since the chosen
node (see :ref:`l2_announcements_leader_election`) will act as the north/south
load balancer for all of the traffic for a particular service.

Interfaces
----------

> The interfaces field is a list of regular expressions (`golang syntax <https://pkg.go.dev/regexp/syntax>`__

Reviewer 3: ```suggestion
The interfaces field is a list of regular expressions (`golang syntax <https://pkg.go.dev/regexp/syntax>`__)
```

that determine over which network interfaces the selected services will be 
announced. This field is optional, if not specified all interfaces will be used.

The expressions are OR-ed together, so any network device matching any of the 
expressions will be matched.

L2 announcements only work if the selected devices are also part of the set of 

> devices specified in the ``devices`` Helm option, see :ref:`NodePort Devices`

Reviewer 3: ```suggestion
devices specified in the ``devices`` Helm option, see :ref:`NodePort Devices`.
```

.. code-block:: shell-session

    $ kubectl -n kube-system get lease/cilium-l2announce-default-deathstar -o yaml
    apiVersion: coordination.k8s.io/v1
    kind: Lease
    metadata:
      creationTimestamp: "2023-05-09T15:13:32Z"
      name: cilium-l2announce-default-deathstar
      namespace: kube-system
      resourceVersion: "449966"
      uid: e3c9c020-6e24-4c5c-9df9-d0c50f6c4cec
    spec:
      acquireTime: "2023-05-09T15:14:20.108431Z"
      holderIdentity: worker-node
      leaseDurationSeconds: 3
      leaseTransitions: 1
      renewTime: "2023-05-12T12:15:26.773020Z"

The ``acquireTime`` is the time at which the current leader acquired the lease.
The ``holderIdentity`` is the name of the current holder/leader node. 
If the leader does not renew the lease for ``leaseDurationSeconds`` seconds a
new leader is chosen. ``leaseTransitions`` indicates how often the lease changed
hands and ``renewTime`` the last time the leader renewed the lease.

There are three Helm options that can be tuned with regards to leases:

* ``l2announcements.leaseDuration`` determines the ``leaseDurationSeconds`` value
  of created leases and by extent how long a leader must be "down" before 

>   failover occurs. Its default value is 15s, it must always be greater then 1s

Reviewer 3: `then` -> `than`
